farm/train.py

```diff
@@ -160,12 +247,36 @@ def train(self, model):
                     if self.global_step != 0 and (
                         self.global_step % self.evaluate_every == 0
                     ):
+                        evalnr += 1
                         result = self.evaluator_dev.eval(model)
                         self.evaluator_dev.log_results(result, "Dev", self.global_step)
-
+                        if self.early_stopping:
+                            do_stopping, save_model, eval_value = self.early_stopping.check_stopping(result)
+                            if save_model:
+                                logger.info(
+                                    "Saving current best model to {}, eval={}".format(
+                                        self.early_stopping.save_dir, eval_value))
+                                model.save(self.early_stopping.save_dir)
+                                self.data_silo.processor.save(self.early_stopping.save_dir)
+                            if do_stopping:
+                                # log the stopping
+                                logger.info("STOPPING EARLY AT EPOCH {}, STEP {}, EVALUATION {}".format(epoch, step, evalnr))
+                if do_stopping:
+                    break
                 self.global_step += 1
-
+            if do_stopping:
+                break
         if self.evaluator_test is not None:
+            if self.early_stopping and self.early_stopping.save_dir:
+                logger.info("Restoring best model so far from {}".format(self.early_stopping.save_dir))
+                # if we have early stopping and saved a model, re-store it now to the same device
+                # We are getting an exception when doing this
+                model_device_type = next(iter(model.parameters())).device.type
```

Comment:
No, that is partly because I initially was not sure how flexible we need to be about the kind of model to expect here and partly because I copied it from a location where self.device did not exist. I am changing this to what you suggested.